core/trino-main/src/main/java/io/trino/operator/PageBuffer.java

```diff
@@ -98,4 +98,9 @@ public void finish()
     {
         finished = true;
     }
+
+    public void unfinish()
+    {
+        finished = false;
+    }
```

Comment:
This will break how processors and state in general is intended to work in Trino. Once a a Processor moves to a finished state, it shouldn't ever be able to move out of it. It is expected to be a terminal state.

Can you explain a bit more about what you are trying to do with this?

Reply:
This is due to the way WorkProcessor works. TopNPages.process will be never called until TopNOperator is finishing(a.k.a. pageBuffer.finish is called). So, flushing is not possible until TopNOperator's state is changed to finishing. If the state is not changed back from finishing(pageBuffer.finished == true) after the flushing, TopNOperator cannot receive the rest input pages. So, the rest input pages will be discarded. TopNRowNumberOperator and HashAggregationOperator haven't been migrated into WorkProcessor. So, those operators' flushing implementations don't have this issue. I know this might be a bit tricky. Please let me know if you have a better idea. Feel free to ping me through Slack as well.

Reply:
I changed the logic to utilize PageBuffer, instead.